File: flutter/app/lib/src/flow/use_smileid_sample_flow_launch_snapshot.dart
import 'package:flutter/foundation.dart';
import 'package:sample_ui/sample_ui.dart';

import 'use_smileid_sample_token_binding_rules.dart';

/// Read once at flow entry; never re-read while the flow runs.
@immutable
class UseSmileIDSampleFlowLaunchSnapshot {
  /// Takes the five settings rather than the settings object, because the read happens once.
  const UseSmileIDSampleFlowLaunchSnapshot({
    required this.product,
    required this.route,
    required this.userDetails,
    required this.idDetails,
    required this.scenario,
    required this.theme,
    required this.sandbox,
    required this.allowAgentMode,
    required this.enableEnhancedLiveness,
    required this.consentStep,
    required this.instructionsStep,
    required this.previewStep,
    required this.userId,
    required this.partnerId,
    required this.partnerName,
    required this.callbackUrl,
    this.session,
    this.sessionExpired = false,
  });

  /// The product this run submits.
  final UseSmileIDSampleProduct product;

  /// Which presentation hosts the flow.
  final UseSmileIDSampleFlowRoute route;

  /// What the user-details form collected.
  final UseSmileIDSampleUserDetails userDetails;

  /// What the ID form collected.
  final UseSmileIDSampleIdDetails idDetails;

  /// The flow scenario in effect.
  final UseSmileIDSampleScenario scenario;

  /// The theme scenario in effect.
  final UseSmileIDSampleThemeScenario theme;

  /// Whether the run submits against sandbox.
  final bool sandbox;

  /// Whether an operator may capture for the applicant.
  final bool allowAgentMode;

  /// Whether the head-turn challenge runs.
  final bool enableEnhancedLiveness;

  /// Whether the SDK's consent step is composed in.
  final bool consentStep;

  /// Whether the instructions step is composed in.
  final bool instructionsStep;

  /// Whether a preview follows each capture.
  final bool previewStep;

  /// The id this run submits under.
  final String userId;

  /// The active profile's id, which the SDK submits as the partner.
  final String partnerId;

  /// The active profile's organisation, which the consent screen names.
  final String partnerName;

  /// The active profile's webhook URL; empty means their portal default.
  final String callbackUrl;

  /// Live at entry only: a session that has run out is the gate's business, never the builder's.
  final UseSmileIDSampleTokenSession? session;

  /// Run out, which is the one thing that routes back to the scanner; usually true with no [session].
  final bool sessionExpired;

  /// The session a run submits under, absent for the two scenarios that are about refresh.
  UseSmileIDSampleTokenSession? get liveSession =>
      useSmileIDSampleStartsExpired(scenario) ? null : session;

  /// Where the run submitted, which is the only thing that publishes it.
  UseSmileIDSampleEnvironment get environment => sandbox
      ? UseSmileIDSampleEnvironment.sandbox
      : UseSmileIDSampleEnvironment.production;
}

File: expo/app/src/flow/use-smile-id-sample-flow-launch-snapshot.ts
import type {
  UseSmileIDSampleFlowRoute,
  UseSmileIDSampleIdDetails,
  UseSmileIDSampleProduct,
  UseSmileIDSampleTokenSession,
  UseSmileIDSampleUserDetails,
} from '@smileid/sample-ui';

import { smileIDSampleStartsExpired } from './use-smile-id-sample-token-binding-rules';

/// Read once at flow entry; never re-read while the flow runs.
export type UseSmileIDSampleFlowLaunchSnapshot = {
  readonly product: UseSmileIDSampleProduct;
  readonly route: UseSmileIDSampleFlowRoute;
  readonly userDetails: UseSmileIDSampleUserDetails;
  readonly idDetails: UseSmileIDSampleIdDetails;
  /// The scenario ids, which on this platform arrive by launch argument rather than from a drawer.
  readonly scenario: string;
  readonly theme: string;
  /// Whether the run submits against sandbox; the linked session decides, and no session is sandbox.
  readonly sandbox: boolean;
  /// The five settings, not the settings object: the read happens once.
  readonly allowAgentMode: boolean;
  readonly enableEnhancedLiveness: boolean;
  readonly consentStep: boolean;
  readonly instructionsStep: boolean;
  readonly previewStep: boolean;
  /// The id this run submits under; only authentication sends it.
  readonly userId: string;
  readonly partnerId: string;
  readonly partnerName: string;
  /// The active profile's webhook URL; empty means their portal default.
  readonly callbackUrl: string;
  /// Live at entry only: a session that has run out is the gate's business, never the builder's.
  readonly session: UseSmileIDSampleTokenSession | null;
  /// Run out — the one thing that routes back to the scanner. Usually true with no session.
  readonly sessionExpired: boolean;
};

/// Where the run submitted, which is the only thing that publishes it.
export const smileIDSampleFlowEnvironment = (
  snapshot: UseSmileIDSampleFlowLaunchSnapshot,
): 'sandbox' | 'production' => (snapshot.sandbox ? 'sandbox' : 'production');

/// The session a run actually submits under.
export const smileIDSampleSnapshotSession = (
  snapshot: UseSmileIDSampleFlowLaunchSnapshot,
): UseSmileIDSampleTokenSession | null =>
  smileIDSampleStartsExpired(snapshot.scenario) ? null : snapshot.session;
